libimage/disk_usage.go

```diff
@@ -155,3 +145,23 @@ func diskUsageForImage(ctx context.Context, image *Image, tree *layerTree) ([]Im
 
 	return results, nil
 }
+
+// walkImageLayers walks all layers in an image and calls the given function for each layer.
+func walkImageLayers(image *Image, layerMap map[string]*storage.Layer, f func(layer *storage.Layer)) {
+	// Layers are walked recursively until it has no parent which means we reached the end.
+	// We must account for the fact that an image might have several top layers when id mappings are used.
+	layers := append([]string{image.storageImage.TopLayer}, image.storageImage.MappedTopLayers...)
+	for _, layerID := range layers {
+		for layerID != "" {
+			layer := layerMap[layerID]
+			if layer == nil {
+				logrus.Errorf("Local Storage is corrupt, layer %q missing from the storage", layerID)
+				break
+			}
+			f(layer)
+			// Set the layer for the next iteration, parent is empty if we reach the end.
+			// FIXME: This assumes there are no loops, is that a guarantee c/storage gives us?
```

Comment:
By design, ~yes: a layer’s parent can only be set to an existing layer, and it can’t be changed afterwards.

But I don’t think the code protects against invalid data stored on disk, or against additional layer stores violating this, for example.